allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+        # last one.
+        assert token_answer_span_end > token_answer_span_start
+
+        # Tokenize the question
+        tokenized_question = self._tokenizer.tokenize(question)
+        tokenized_question = tokenized_question[: self.max_query_length]
+        tokenized_question = [
+            token._replace(type_id=self.non_content_type_id, idx=None)
+            for token in tokenized_question
+        ]
+
+        # Stride over the context, making instances
+        # Sequences are <s> question </s> </s> question </s>.
+        # This is technically not correct for anything but RoBERTa, but it does not affect the scores.
+        space_for_context = self.length_limit - len(tokenized_question) - 4
+        stride_start = 0
+        while True:
+            tokenized_context_window = tokenized_context[stride_start:]
+            tokenized_context_window = tokenized_context_window[:space_for_context]
+
+            window_token_answer_span = (
+                token_answer_span_start - stride_start,
+                token_answer_span_end - stride_start,
+            )
+            if any(i < 0 or i > len(tokenized_context_window) for i in window_token_answer_span):
+                # The answer is not contained in the window.
+                window_token_answer_span = None
+
+            if not self.skip_invalid_examples or window_token_answer_span is not None:
+                additional_metadata = {"id": qid}
+                instance = self.text_to_instance(
+                    question,
+                    tokenized_question,
+                    context,
+                    tokenized_context_window,
+                    answers,
+                    window_token_answer_span,
+                    additional_metadata,
+                )
+                yield instance
+
+            stride_start += space_for_context
+            if stride_start >= len(tokenized_context):
+                break
+            stride_start -= self.stride
+
+    @overrides
+    def text_to_instance(
+        self,  # type: ignore
+        question: str,
+        tokenized_question: List[Token],
+        context: str,
+        tokenized_context: List[Token],
+        answers: List[str],
+        token_answer_span: Optional[Tuple[int, int]],
+        additional_metadata: Dict[str, Any] = None,
+    ) -> Instance:
+        fields = {}
+
+        # make the question field
+        def cls_token():
+            return Token(
+                self._tokenizer.tokenizer.cls_token,
+                text_id=self._tokenizer.tokenizer.cls_token_id,
+                type_id=self.non_content_type_id,
+            )
+
+        def sep_token():
+            return Token(
+                self._tokenizer.tokenizer.sep_token,
+                text_id=self._tokenizer.tokenizer.sep_token_id,
+                type_id=self.non_content_type_id,
+            )
+
+        question_field = TextField(
```

Comment:
Is this generalizable? I'm doing it this way: https://github.com/allenai/allennlp/blob/6babeef119ad02ed184e265a57b1df563a0c3376/allennlp/data/tokenizers/pretrained_transformer_tokenizer.py#L201-L213

Reply:
I saw your approach, and I tried it here, because it bothers me that we're using the RoBERTa special token scheme for all the transformers. But doing this properly it gets complicated because of the sliding windows. Since the scores were the same anyways, I decided not to do it.

Reply:
Is it the `4` in `space_for_context = self.length_limit - len(tokenized_question) - 4` that is problematic, or something else?

Reply:
Yeah, I don't really get why the sliding window makes a difference here.  (I don't have a strong opinion about whether to hard-code two SEP tokens or not; it doesn't seem like a big deal either way.  Just curious.)

Reply:
I think the story goes like this: Huggingface's `tokenizer.encode()` does everything. It makes tokens, it makes token ids, it calculates sliding windows, and it adds special tokens. It does not create character offsets however, so we use our own tokenizer for this. If we want to delegate the rest of the work to the huggingface library, we have to map our token representation back to theirs, let it do the windowing and the special characters, and then map it back into our representation. This seems way more complicated than being a little bit wrong for non-RoBERTa models, especially given that the scores are the same.

Reply:
Okay. I also realized that the `intra_word_tokenize()`, `num_added_start_tokens` and `num_added_end_tokens` currently only apply to single sequence input but not sequence pairs. But it's just been merged if it could be useful in any way.

Reply:
Ah, got it. To rephrase to be sure I understand: it's complicated to do it the strictly correct way because it means we need to add a lot of logic that we would otherwise just get from huggingface.  Seems like a fine tradeoff to me.

Reply:
We might look at this again once we have the new tokenizers, the ones that already have character offsets. Then we could potentially delegate all of it to huggingface.